src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -2286,13 +2294,23 @@ int CpuemShim::xrtGraphTimedWait(void * gh, uint64_t cycle) {
 * forever or graph that has multi-rate core(s).
 */
 int CpuemShim::xrtGraphEnd(void * gh) {
-  bool ack = false;
+
+  uint32_t ack = false;
+
   auto ghPtr = (xclcpuemhal2::GraphType*)gh;
   if (!ghPtr)
     return -1;
+
   auto graphhandle = ghPtr->getGraphHandle();
-  xclGraphEnd_RPC_CALL(xclGraphEnd, graphhandle);
-  if (!ack)
+  do
+  {
+    mApiMtx.lock();
+    xclGraphEnd_RPC_CALL(xclGraphEnd, graphhandle);
+    mApiMtx.unlock();
```

Comment:
Use rather RAII instead of dangerous raw low-level API https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#i7-state-postconditions
```suggestion
    {
      std::lock_guard lk { mApiMtx };
      xclGraphEnd_RPC_CALL(xclGraphEnd, graphhandle);
    }
```

Reply:
I have added your suggestion in the latest commit.